<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Document</title>
</head>
<body>
  <script>

    class yPromise {
      static PENDING = 'pending'
      static FULFILLED = 'fulfilled'
      static REJECTED = 'rejected'
      constructor(executor) {
        // 初始化状态
        this.status = yPromise.PENDING
        // 储存操作成功返回的值
        this.value = undefined
        // 储存操作拒绝返回的原因
        this.reason = undefined
        // 储存then中的参数
        // 至于为什么是数组呢？因为同一个Promise的then方法可以调用多次
        this.callbacks = []
        // 绑定this为了解决指向问题
        executor(this.resolve.bind(this), this.reject.bind(this))
      }

      // onFulfilled成功执行的函数
      // onRejected失败执行的函数
      then(onFulfilled, onRejected) {
        // 每一次then都应该返回一个new promise 确保链式调用
        return new yPromise((nextResolve, nextReject) => {
          // this.callbacks.push({
          //   onFulfilled,
          //   onRejected
          // })
          this.handler({
            nextResolve,
            nextReject,
            onFulfilled,
            onRejected
          })
        })
      }
      // catch只有一个参数用于处理错误的情况
      catch(onRejected) {
        return this.then(null, onRejected)
    }

      // 成功失败都会执行的方法
      finally(onFinally) {
        return this.then(onFinally, onFinally)
      }

      // 成功后的回调信息
      resolve(value) {
        // 如果传过来的是promise 说明不是第一次调用
        if (value instanceof yPromise) {
          value.then(
            this.resolve.bind(this),
            this.reject.bind(this)
          )
        }
        this.value = value
        // 将状态设置为成功
        this.status = yPromise.FULFILLED
        // 通知事件执行
        this.callbacks.forEach(cb => this.handler(cb))
      }

      // 拒绝后的回调信息
      reject(reason) {
        this.reason = reason
        // 将状态设置为失败
        this.status = yPromise.REJECTED
        // 通知事件执行
        this.callbacks.forEach(cb => this.handler(cb))
      }

      handler(callback) {
        // 获取回调函数中的方法
        const { onFulfilled, onRejected, nextResolve, nextReject } = callback

        if (this.status === yPromise.PENDING) {
          this.callbacks.push(callback)
          return
        }

        if (this.status === yPromise.FULFILLED) {
          const nextValue = onFulfilled ? onFulfilled(this.value) : undefined
          nextResolve(nextValue)
          return
        }

        if (this.status === yPromise.REJECTED) {
          const nextReason = onRejected ? onRejected(this.reason) : undefined
          nextReject(nextReason)
        }
      }
    }

    function fetchData() {
      return new yPromise((resolve, reject) => {
        setTimeout(() => {
          reject(1)
        })
      })
    }

    fetchData().then(res => {
      console.log(res)
    }).catch((e) => {
      console.log(e)
      console.log('失败了')
    })

    function test() {
      return new Promise((resolve, reject) => {
        reject(1)
      })
    }
    test().then(res => {
      console.log(res)
    }).catch(err => {
      console.log(err)
    })
    // test().then(res => {
    //   console.log(res)
    //   return new Promise((resolve, reject) => {
    //     resolve(2)
    //   })
    // }).finally( err => {
    //   console.log(1)
    // })
  </script>
</body>
</html>